pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,389 @@
+		return fmt.Errorf("could not create directory %s for new update: %w", newUpdateDirectory, err)
+	}
+
+	removeBrokenUpdateDir := func() {
+		if err := os.RemoveAll(newUpdateDirectory); err != nil {
+			level.Debug(ulm.logger).Log(
+				"msg", "could not remove broken update directory",
+				"update_dir", newUpdateDirectory,
+				"err", err,
+			)
+		}
+	}
+
+	if err := fsutil.UntarBundle(filepath.Join(newUpdateDirectory, string(binary)), stagedUpdate); err != nil {
+		removeBrokenUpdateDir()
+		return fmt.Errorf("could not untar update to %s: %w", newUpdateDirectory, err)
+	}
+
+	// Make sure that the binary is executable
+	if err := os.Chmod(executableLocation(newUpdateDirectory, binary), 0755); err != nil {
+		removeBrokenUpdateDir()
+		return fmt.Errorf("could not set +x permissions on executable: %w", err)
+	}
+
+	// Validate the executable
+	if err := autoupdate.CheckExecutable(context.TODO(), executableLocation(newUpdateDirectory, binary), "--version"); err != nil {
+		removeBrokenUpdateDir()
+		return fmt.Errorf("could not verify executable: %w", err)
+	}
+
+	return nil
+}
+
+// currentRunningVersion returns the current running version of the given binary.
+func (ulm *updateLibraryManager) currentRunningVersion(binary autoupdatableBinary) (*semver.Version, error) {
+	switch binary {
+	case binaryLauncher:
+		currentVersion, err := semver.NewVersion(version.Version().Version)
+		if err != nil {
+			return nil, fmt.Errorf("cannot determine current running version of launcher: %w", err)
+		}
+		return currentVersion, nil
+	case binaryOsqueryd:
+		resp, err := ulm.osquerier.Query("SELECT version FROM osquery_info;")
+		if err != nil {
+			return nil, fmt.Errorf("could not query for osquery version: %w", err)
+		}
+		if len(resp) < 1 {
+			return nil, errors.New("osquery version query returned no rows")
+		}
+		rawVersion, ok := resp[0]["version"]
+		if !ok {
+			return nil, errors.New("osquery version query did not return version")
+		}
+
+		currentVersion, err := semver.NewVersion(rawVersion)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse current running version %s of osquery as semver: %w", rawVersion, err)
+		}
+
+		return currentVersion, nil
+	default:
+		return nil, fmt.Errorf("cannot determine current running version for unexpected binary %s", binary)
+	}
+}
+
+// TidyLibrary removes unneeded files from the staged updates and updates directories.
+func (ulm *updateLibraryManager) TidyLibrary() {
+	for _, binary := range binaries {
+		// First, remove old staged archives
+		ulm.tidyStagedUpdates(binary)
+
+		// Get the current running version to preserve it when tidying the available updates
+		var currentVersion *semver.Version
+		var err error
+		switch binary {
+		case binaryOsqueryd:
+			// The osqueryd client may not have initialized yet, so retry the version
+			// check a couple times before giving up
+			osquerydVersionCheckRetries := 5
+			for i := 0; i < osquerydVersionCheckRetries; i += 1 {
```

Comment:
Currently, all of this is ultimately run by a `rungroup.Run` (Run => TufAutoupdater.Execute => updateLibraryManager.TidyLibrary), so yeah, within a goroutine.

We did discuss doing the tidying on launcher startup (after version selection), which sounds like it would be blocking in that case. Should we talk through that idea more? I am also fine with execing it with --version. 🤷‍♀️

Reply:
I don't think we know the simple answer, so I might be inclined to solve it later.

With the initial delay on autoupdater, I think we don't have an immediate problem. But it feels like it's not right.